File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        # head alone edge case
        if not head or not head.next:
            return None

        # get the size of the linked list
        l = 0
        p = head
        while p:
            l += 1
            p = p.next

        # l == n edge case
        if l==n:
            return head.next
        
        # remove the nth node from the end
        p = head
        q = p
        for _ in range(l - n):
            q = p
            p = p.next
        q.next = p.next

        return head

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode* p=head;
    struct ListNode* q=head;
    int len=0;

    // head alone edge case
    if (!head || !head->next)return NULL;

    // get the length of the linked list
    for (;p;p=p->next)len++;

    // len == n edge case
    if (len==n)return head->next;

    // remove nth node from the end
    p=head;
    for (int i=0; i<len-n; i++){
        q=p;
        p=p->next;
    }
    q->next=p->next;

    return head;
}